### PR TITLE
Add lapack95.

### PR DESCRIPTION
**Pull request by hhslepicka**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Completely based on https://anaconda.org/birdhouse/lapack95

### CODE REVIEW

**Comment by hhslepicka** [comment]
@conda-forge/staged-recipes this one is ready to go. Thank you!

**Comment by jakirkham** [comment]
We have some [existing specification]( https://conda-forge.org/docs/maintainer/knowledge_base.html#blas ) around how BLAS/LAPACK are handled in conda-forge. I think we would want this to build on top of that work somehow, but am not familiar enough with this library or how it is used to advise on how best to do this. Maybe as a first step would recommend reading up on those docs and then sharing a bit more about how this is intended to be used

**Comment by hhslepicka** [comment]
@jakirkham this library is a Fortran95 interface to LAPACK.
I submitted this PR to add it into conda-forge as another package that I am working on depends on a file provided by this package (`f95_lapack.mod`).

File: recipes/lapack95/meta.yaml
Comment (jakirkham) [inline recipes/lapack95/meta.yaml:23]: This shouldn’t be needed as the Fortran compiler will already add it’s runtime libraries as dependencies

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: Addressed via https://github.com/conda-forge/staged-recipes/pull/14212/commits/02def9f4d92b182dbd2ed82c4db29bbdf4d25d59

File: recipes/lapack95/make.inc
Comment (jakirkham) [inline recipes/lapack95/make.inc:28]: Is it possible to produce dynamic libraries instead?

File: recipes/lapack95/meta.yaml
Comment (jakirkham) [inline recipes/lapack95/meta.yaml:9]: Do they provide a versioned URL or do we know of an FTP server that does? That would make tracking and updating to new releases easier

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml:9]: That is the official website for it... unfortunately other sources that claim mirroring of it modified the code a lot to the point I am not sure if it would produce the same result.

File: recipes/lapack95/make.inc
Comment (hhslepicka) [reply recipes/lapack95/make.inc:28]: I will try to take a look into it. And get back to you on that. Should I look into provide both?

Comment (jakirkham) [reply recipes/lapack95/make.inc:28]: Ok thanks. Just the dynamic libraries are fine

Maintainers can optionally provide static libraries in a separate package, but that’s not needed

This CFEP is relevant

https://github.com/conda-forge/cfep/blob/master/cfep-18.md

File: recipes/lapack95/meta.yaml
Comment (jakirkham) [reply recipes/lapack95/meta.yaml:9]: Ah ok. How do we know if a new version is released?

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml:9]: Probably from periodic visits to their website. Latest release which is this one is from `Date:  November 30, 2000`.

File: recipes/lapack95/make.inc
Comment (hhslepicka) [reply recipes/lapack95/make.inc:28]: I looked over the instructions provided with the package there are no existing ways to build a dynamic version. Can we go with the static one? I can try later to patch their build system and try to get a shared version.

**Comment by hhslepicka** [comment]
@jakirkham can we get this one in? I need it to move forward with #14234 .

File: recipes/lapack95/make.inc
Comment (isuruf) [reply recipes/lapack95/make.inc:28]: This is not creating a static library. Here you are using the static versions of the dependencies. Please use the dynamic libraries of dependencies

**Review by isuruf — CHANGES_REQUESTED** [review]

File: recipes/lapack95/make.inc
Comment (hhslepicka) [reply recipes/lapack95/make.inc:28]: @isuruf it creates `lapack95.a` (static) and that is hardcoded all over the makefile for this library.
I can change the dependencies to be shared and see how that works.

Comment (isuruf) [reply recipes/lapack95/make.inc:28]: That's not what I was talking about. For eg:  `BLAS     = $(LAPACK_PATH)/libblas.a` uses `libblas.a` dependency from `libblas` package in conda-forge.

Comment (hhslepicka) [reply recipes/lapack95/make.inc]: @isuruf sorry for the misunderstanding. Addressed via https://github.com/conda-forge/staged-recipes/pull/14212/commits/cfe495c7a35765a0cfda06f19aee443b0b6bbfdd

File: recipes/lapack95/make.inc
Comment (isuruf) [inline recipes/lapack95/make.inc:26]:
```suggestion
LAPACK77 = $(LAPACK_PATH)/liblapack$(SHLIB_EXT)
```

Comment (hhslepicka) [reply recipes/lapack95/make.inc]: Addressed via https://github.com/conda-forge/staged-recipes/pull/14212/commits/38b23a22c79870d6858d239a949e8b6f7a6d0b3b for LAPACK and BLAS.

File: recipes/lapack95/meta.yaml
Comment (jakirkham) [inline recipes/lapack95/meta.yaml:25]:
```suggestion
    - test -f $PREFIX/lib/liblapack95$SHLIB_EXT
```

Comment (jakirkham) [reply recipes/lapack95/meta.yaml:25]: Also the static library is not being produced right? Might be worth adding a test that ensure the static library is not packaged

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: @jakirkham the static library is being produced and packaged.

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: I mentioned here https://github.com/conda-forge/staged-recipes/pull/14212#discussion_r592581467 that the compiling of the static library is hardcoded on the makefile for this package. As requested by @isuruf I changed to use the shared libraries as dependencies but I am not comfortable with rewriting the makefile for this library to have a shared library at least not at this moment if that would be ok.

File: recipes/lapack95/build.sh
Comment (jakirkham) [inline recipes/lapack95/build.sh:9]:
```suggestion
cp ../lapack95$SHLIB_EXT $PREFIX/lib/liblapack95$SHLIB_EXT
```

File: recipes/lapack95/make.inc
Comment (jakirkham) [inline recipes/lapack95/make.inc:25]:
```suggestion
LAPACK95 = ../lapack95$SHLIB_EXT
```

Comment (hhslepicka) [reply recipes/lapack95/make.inc:25]: `SHLIB_EXT` will make it be `.so`, right? If yes, I can't do it as the makefile has the hardcoded name to generate the file as `lapack95.a`.
Here:
image

Comment (jakirkham) [reply recipes/lapack95/make.inc:25]: We can add and apply a patch to fix that

https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#patches

Comment (hhslepicka) [reply recipes/lapack95/make.inc:25]: I am familiar with the possibility of applying patches to packages I don't think it is right to name a static library as `x.so`.
Is it the case that unless I rewrite their makefile or patch it to produce a shared library this recipe won't be merged?

If that is the only solution that's fine but I would like to avoid going this route if possible. Please don't get me wrong it is just that this is a library that no one touched since November 2000 and I see little to no benefit on the change other than the reasoning from CFEP-18.

Let me know how you would like me to proceed so this can be merged.

Comment (isuruf) [reply recipes/lapack95/make.inc:25]: @hhslepicka, you can create a shared library from static library without having to rewrite the makefile.
One problem with static builds would be that if you use this package in a downstream package, then you'll have to package the license of this package in the downstream package which makes it messy.

Comment (hhslepicka) [reply recipes/lapack95/make.inc:25]: @isuruf could you share how to do this (create the shared from the static) without rewriting the makefile?

Comment (isuruf) [reply recipes/lapack95/make.inc:25]: after building the static library, do the following,
```bash
if [[ "$target_platform" == linux-* ]]; then
  $FC  -shared -o -Wl,--whole-archive  liblapack95.so  -Wl,--no-whole-archive $PREFIX/lib/liblapack.so $PREFIX/lib/libblas.so 
elif [[ "$target_platform" == osx-* ]]; then
  $FC -shared -o  -Wl,-all_load  liblapack95.dylib -Wl,-noall_load $PREFIX/lib/liblapack.dylib $PREFIX/lib/libblas.dylib 
fi
```
It's untested, so there might be typos above

Comment (hhslepicka) [reply recipes/lapack95/make.inc:25]: I will give it a try here and get back to you on that...

Comment (hhslepicka) [reply recipes/lapack95/make.inc:25]: @isuruf I am not getting where the static lib goes into play on the generation of the shared one there. It does not use `liblapack95.a` anywhere.

**Comment by hhslepicka** [comment]
@isuruf I managed to put a patch to change the makefile and build a proper shared library that also works with my other package that I need it for.
The issue that I have now is that somehow for osx it is failing as it can't find some symbols that are supposed to come from LAPACK.
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=289392&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1482

For linux it is working fine. Also a local build at my osx machine using the code + patch + packages in the recipe results in a successful build.

Do you mind taking a look?

**Comment by hhslepicka** [comment]
@isuruf @jakirkham Something weird is going on with osx build.
Here (https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=289623&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=933) it says that liblapack and libblas will be installed but when I do some debugging via `conda list` (https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=289623&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1471) I don't seem them on the list which makes sense when it fails to link the shared library.

Do you mind taking a look?

File: recipes/lapack95/build.sh
Comment (isuruf) [inline recipes/lapack95/build.sh:16]:
```suggestion
  conda list -p $PREFIX
```

**Comment by hhslepicka** [comment]
@isuruf a difference that I just noticed is that locally I am testing with liblapack 3.9 - build 8_openblas.

```$conda list
# packages in environment at /Users/slepicka/mc/envs/lapack95-test:
#
# Name                    Version                   Build  Channel
cctools                   949.0.1             hd9211c8_20    conda-forge
cctools_osx-64            949.0.1             h6407bdd_20    conda-forge
clang                     11.1.0               h694c41f_0    conda-forge
clang-11                  11.1.0          default_he082bbe_0    conda-forge
clang_osx-64              11.1.0               hb91bd55_0    conda-forge
clangxx                   11.1.0          default_he082bbe_0    conda-forge
compiler-rt               11.1.0               h654b07c_0    conda-forge
compiler-rt_osx-64        11.1.0               h8c5fa43_0    conda-forge
fortran-compiler          1.1.3                h6818c3e_0    conda-forge
gfortran_impl_osx-64      9.3.0               h9cc0e5e_19    conda-forge
gfortran_osx-64           9.3.0               h18f7dce_13    conda-forge
gmp                       6.2.1                h2e338ed_0    conda-forge
isl                       0.22.1               hb1e8313_2    conda-forge
ld64                      530                 hd2e7500_20    conda-forge
ld64_osx-64               530                 he8994da_20    conda-forge
ldid                      2.1.2                h7660a38_2    conda-forge
libblas                   3.9.0                8_openblas    conda-forge
libclang-cpp11.1          11.1.0          default_he082bbe_0    conda-forge
libcxx                    11.1.0               habf9029_0    conda-forge
libgfortran               5.0.0           9_3_0_h6c81a4c_19    conda-forge
libgfortran-devel_osx-64  9.3.0               h6c81a4c_19    conda-forge
libgfortran5              9.3.0               h6c81a4c_19    conda-forge
libiconv                  1.16                 haf1e3a3_0    conda-forge
liblapack                 3.9.0                8_openblas    conda-forge
libllvm11                 11.1.0               hd011deb_0    conda-forge
libopenblas               0.3.12          openmp_h54245bb_1    conda-forge
llvm-openmp               11.0.1               h7c73e74_0    conda-forge
llvm-tools                11.1.0               hd011deb_0    conda-forge
make                      4.3                  h22f3db7_1    conda-forge
mpc                       1.1.0             ha57cd0f_1009    conda-forge
mpfr                      4.0.2                h72d8aaf_1    conda-forge
tapi                      1100.0.11            h9ce4665_0    conda-forge
zlib                      1.2.11            h7795811_1010    conda-forge
```

**Comment by hhslepicka** [comment]
@isuruf @jakirkham this is good to go if you agree with my approach to `openblas`.

File: recipes/lapack95/meta.yaml
Comment (isuruf) [inline recipes/lapack95/meta.yaml:24]:
```suggestion
  host:
    - libblas
    - liblapack
```

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml:24]: @isuruf if I remove openblas it will install the `netlib` alternative of libblas and therefore I will go back into the issue that I had from the missing symbols. 
More details here: https://github.com/conda-forge/staged-recipes/pull/14212#issuecomment-797701592

Comment (isuruf) [reply recipes/lapack95/meta.yaml:24]: which symbols are missing?

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml:24]: `_clagge_` is one of them... I can apply your patch and get a full list.

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml:24]: If you try this:
`nm $PREFIX/lib/liblapack.dylib | grep clagge` on a MacOS machine with lapack from netlib or blis it won't work.. but it does for mkl and openblas.

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: I pushed a change so it can run the CI without openblas and all the missing symbols will show. My debug will also print the output of `nm` for liblapack.dylib to confirm the missing symbols.

Comment (isuruf) [reply recipes/lapack95/meta.yaml]: Ah, the testing symbols are missing. Let me fix that.

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: Awesome! Thanks

Comment (hhslepicka) [reply recipes/lapack95/meta.yaml]: @isuruf here is the full list of missing symbols:
```
Undefined symbols for architecture x86_64:
  "_clagge_", referenced from:
      _clagge_f95_ in lapack95.a(la_clagge.o)
     (maybe you meant: _clagge_f95_)
  "_dlagge_", referenced from:
      _dlagge_f95_ in lapack95.a(la_dlagge.o)
     (maybe you meant: _dlagge_f95_)
  "_slagge_", referenced from:
      _slagge_f95_ in lapack95.a(la_slagge.o)
     (maybe you meant: _slagge_f95_)
  "_zlagge_", referenced from:
      _zlagge_f95_ in lapack95.a(la_zlagge.o)
     (maybe you meant: _zlagge_f95_)
ld: symbol(s) not found for architecture x86_64
```


More details, including `nm liblapack.dylib` here:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=290081&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=3595

**Comment by hhslepicka** [comment]
@isuruf thank you for adding the flags for dynamic lookup.
I cleaned up the debug prints.

**Pull request merged**
